### PR TITLE
BENCH: Add benchmarks for stats module

### PR DESCRIPTION
**Pull request by maniteja123**
Add some benchmarks for stats package
- [x] sample test for fischer_exact
- [x] similar benchmarks for other functions like ttest_ind
- [x] also one distribution with its methods(pdf, cdf and fit being the most important), maybe cauchy, gamma or beta distribution.

~~As of now, it is unable to run this file using `python runtests.py --bench stats` , which gives `No benchmarks selected` output. Any help appreciated.~~

### CODE REVIEW

**Comment by pv** [comment]
The benchmark name starts with tes_, but it should not. See the asv documentation.

**Comment by maniteja123** [comment]
@pv Thanks for pointing out the mistake. I changed it to start with ``time_``, was confused with writing the nose test functions.

File: benchmarks/benchmarks/stats.py [hunk 1]
```diff
@@ -3,15 +3,17 @@
 import warnings
 
 import numpy as np
+from numpy.random import rand
 
 try:
     from scipy.stats import anderson_ksamp
+    from scipy.stats import fisher_exact
+    import scipy.stats as stats
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: I would change the 3 imports above to ``from scipy import stats`` and adapt the code below to that.

File: benchmarks/benchmarks/stats.py [hunk 2]
```diff
@@ -3,15 +3,17 @@
 import warnings
 
 import numpy as np
+from numpy.random import rand
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: Better not to do this, but simply use ``np.random.rand`` where needed

File: benchmarks/benchmarks/stats.py [hunk 3]
```diff
@@ -20,3 +22,60 @@ def time_anderson_ksamp(self):
         with warnings.catch_warnings():
             warnings.simplefilter('ignore', UserWarning)
             anderson_ksamp(self.rand)
+
+def random(size):
+    return rand(*size)
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: this is also not needed

File: benchmarks/benchmarks/stats.py [hunk 4]
```diff
@@ -20,3 +22,60 @@ def time_anderson_ksamp(self):
...
+        self.a = a
+
+    def time_fisher_exact(self, alternative):
+        if alternative == 'two-sided':
+            oddsratio, pvalue = fisher_exact(self.a, alternative='two-sided')
+        elif alternative == 'less':
+            oddsratio, pvalue = fisher_exact(self.a, alternative='less')
+        else:
+            oddsratio, pvalue = fisher_exact(self.a, alternative='greater')
+
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: minor PEP8 nitpick: better to have 2 spaces between top-level functions/classes

File: benchmarks/benchmarks/stats.py [hunk 5]
```diff
@@ -20,3 +22,60 @@ def time_anderson_ksamp(self):
...
+        rvs1 = stats.norm.rvs(loc=5,scale=10,size=500)
+        rvs2 = stats.norm.rvs(loc=5,scale=10,size=500)
+        rvs3 = stats.norm.rvs(loc=5, scale=20, size=500)
+        rvs4 = stats.norm.rvs(loc=5, scale=20, size=100)
+        rvs5 = stats.norm.rvs(loc=8, scale=20, size=100)
+        self.a = rvs1
+        self.b = rvs2
+        self.c = rvs3
+        self.d = rvs4
+        self.e = rvs5
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: You can directly do ``self.a = stats.norm.rvs(...`` etc., that saves 5 lines.

File: benchmarks/benchmarks/stats.py [hunk 6]
```diff
@@ -20,3 +22,60 @@ def time_anderson_ksamp(self):
...
+        stats.ttest_ind(self.a, self.b)
+        stats.ttest_ind(self.a, self.b, equal_var=False)
+
+    def time_ttest_ind_diff_var(self):
+        # test with sample with unequal variances:
+        stats.ttest_ind(self.a, self.c)
+        stats.ttest_ind(self.a, self.c, equal_var=False)
+
+    def time_ttest_ind_diff_size(self):
+        # test with sample with different sizes
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: good idea to test with different sizes. i would make the size difference larger though - I expect 500/100 to behave similarly to 500/500. Maybe make it 500/20 or some such ratio?

File: benchmarks/benchmarks/stats.py [hunk 7]
```diff
@@ -20,3 +22,60 @@ def time_anderson_ksamp(self):
...
+        stats.ttest_ind(self.a, self.d)
+        stats.ttest_ind(self.a, self.d, equal_var=False)
+
+    def time_ttest_ind_diff_all(self):
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: this is probably one benchmark too many. I don't expect any difference between using ``loc=5`` and ``loc=8``.

**Comment by rgommers** [comment]
Nice start @maniteja123

File: benchmarks/benchmarks/stats.py (hunk 1 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: Will do the change. Actually there was a merge conflict as some other commit also added ``stats.py`` to the benchmarks.

File: benchmarks/benchmarks/stats.py (hunk 2 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: Yeah, will do it.

File: benchmarks/benchmarks/stats.py (hunk 3 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: I forgot the syntax of ``np.random.rand`` to take multiple dimensions and hence did unpacking. Will use
  
    np.random.rand(2,2)

File: benchmarks/benchmarks/stats.py (hunk 5 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: Yeah will do it, didn't try to optimize then :)

File: benchmarks/benchmarks/stats.py (hunk 7 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: Yeah, right. I just tried to benchmark all the combinations of means and sizes. Will remove this.

File: benchmarks/benchmarks/stats.py (hunk 6 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: Yeah, so the benchmarks would be better for
* different sizes like 500/20 with a larger difference
* different mean (?)
* different variance (?)

File: benchmarks/benchmarks/stats.py (hunk 4 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: thanks, will remember it from next time :)

File: benchmarks/benchmarks/stats.py (hunk 6 above)
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: Sure. I don't think different mean/variance will make a difference, but you might as well make it different.

Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: So, then the function which does benchmarking on different sizes is necessary. For a worst case, I will use different mean and variance. Shall I remove the two functions before this then ?

Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: I'd keep the one for unequal variances - that is a different code path.

Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: Okay, that makes sense. Sorry for the dumb question. Finally, I will put two functions then, each taking samples with different sizes, and one with same variance and other with different variance. Thanks.

**Comment by maniteja123** [comment]
@rgommers Done the changes.

**Comment by maniteja123** [comment]
* Coming to distributions, would it be preferable to generate a random sample of some size and apply the pdf,cdf and other functions of distributions ?
* And as to which distributions, maybe normal, expon, beta or alpha distributions could be used.Any particular preference while choosing them ?

**Comment by rgommers** [comment]
That would be good. Size probably should be something like 1000 (much smaller and you're mostly looking at overhead). 

I previously suggested ``cauchy``, ``gamma`` or ``beta``, but I have no particular preference. As long as it's a continuous one.

**Comment by maniteja123** [comment]
Thanks for the comments. Now for testing the distributions, can you suggest whether to use frozen distributions like  or just use directly like  ``stats.gamma.pdf`` ?

Is it preferable to have something like

    norm.pdf(norm.rvs(size=1000), loc = 1, scale = 4)

or something like
 
   n = norm(loc=1, scale=4)
   n.pdf(norm.rvs(size=1000))

**Comment by rgommers** [comment]
I'd say direct, frozen dists are used less often.

And I would time each method separately, so don't nest them like ``pdf(rvs(``.

**Comment by maniteja123** [comment]
I have added some benchmarks with ``pdf``, ``cdf ``, ``rvs`` and ``fit`` for ``gamma``, ``cauchy`` and ``beta`` distributions. But I am not sure whether the values of the parameters are good enough for benchmarking or if there is some way better to check the bottlenecks.

**Comment by maniteja123** [comment]
Adding to the ones already benchmarked according to suggestion of @rgommers , it would also be great if anyone has an opinion on particular set of functions which are important and need to be benchmarked or if anything needs to be changed in the present set of benchmarks.

File: benchmarks/benchmarks/stats.py [hunk 8]
```diff
@@ -19,4 +18,84 @@ def setup(self, *args):
...
+        a = np.random.rand(2,2)*10
+        self.a = a
+
+    def time_fisher_exact(self, alternative):
+        if alternative == 'two-sided':
+            oddsratio, pvalue = stats.fisher_exact(self.a, alternative='two-sided')
+        elif alternative == 'less':
+            oddsratio, pvalue = stats.fisher_exact(self.a, alternative='less')
+        else:
+            oddsratio, pvalue = stats.fisher_exact(self.a, alternative='greater')
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: Don't know why I didn't make a comment on this the first time around, but: this whole if/elif/else block is just the same as the one-liner:

    oddsratio, pvalue = stats.fisher_exact(self.a, alternative=alternative)

File: benchmarks/benchmarks/stats.py [hunk 9]
```diff
@@ -19,4 +18,84 @@ def setup(self, *args):
...
+
+
+class CorrelationFunctions(Benchmark):
+    param_names = ['alternative']
+    params = [
+        ['two-sided', 'less', 'greater']
+    ]
+
+    def setup(self, mode):
+        a = np.random.rand(2,2)*10
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: spaces around ``*`` here (pep8)

File: benchmarks/benchmarks/stats.py [hunk 10]
```diff
@@ -19,4 +18,84 @@ def setup(self, *args):
...
+        stats.ttest_ind(self.a, self.c)
+        stats.ttest_ind(self.a, self.c, equal_var=False)
+
+
+class Distribution(Benchmark):
+    param_names = ['distribution', 'properties']    
+    params = [
+        ['cauchy', 'gamma', 'beta'],
+        ['pdf', 'cdf', 'rvs', 'fit']
+    ]
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: I quite like how this works. Benchmark output:

               ============== ========= ========= ========== ========
               --                            properties              
               -------------- ---------------------------------------
                distribution     pdf       cdf       rvs       fit   
               ============== ========= ========= ========== ========
                   cauchy      72.14μs   64.87μs   107.18μs   5.88ms 
                   gamma       38.74μs   42.85μs   116.20μs   7.16ms 
                    beta       41.97μs   46.40μs   202.59μs   7.56ms 
               ============== ========= ========= ========== ========

**Comment by rgommers** [comment]
@maniteja123 the distribution benchmarks look fine to me. So a few minor cleanups, and then this can go in.

**Comment by argriffing** [comment]
The `fit` benchmarks are a little strange, but maybe OK. For example they give explicit initial guesses, whereas a good part of the fitting code is devoted to finding such guesses. Also I think the beta distribution has an unbounded maximum log likelihood with this many free parameters.
```python
>>> import numpy as np
>>> from scipy.stats import beta
>>> x = np.linspace(2, 4, num=13)
>>> beta.logpdf(x, a=0.5, b=0.5, loc=2, scale=2)
array([        inf, -0.55191805, -0.85083655, -1.00088885, -1.08583837,
       -1.13064445, -1.14472989, -1.13064445, -1.08583837, -1.00088885,
       -0.85083655, -0.55191805,         inf])
```

**Comment by argriffing** [comment]
I have a question about the runtests.py/run.py/asv system, not specific to this PR.

I was able to run `$ python runtests.py --bench stats.Distribution` from a checked-out `maniteja123/stats-bench` branch, but when I ran `$ python runtests.py --bench-compare master stats.Distribution` from this branch I got errors like:
```
Error running /usr/bin/git checkout -f 4502e8c20448b5733fdc1b5cabd586599d210cf1
              STDOUT -------->
              
              STDERR -------->
              fatal: reference is not a tree: 4502e8c20448b5733fdc1b5cabd586599d210cf1
```
I was pretty diligent about using `$ git clean -dfx` to remove old builds and environments, and I eventually worked around this error by creating a new branch using `$ git checkout -b maniteja123-stats-bench`. I don't understand why `bench-compare` seems to need this extra step but `bench` doesn't.

**Comment by pv** [comment]
bench doesn't need to check out anything, bench-compare needs to. This
should however work with current asv version. Try to clear out the old
repo mirror by `rm -rf benchmarks/scipy`

**Comment by argriffing** [comment]
@pv Thanks, it worked after removing `benchmarks/scipy`. Sorry for the digression.  I think this PR should be OK to merge after making the changes suggested by @rgommers.

File: benchmarks/benchmarks/stats.py (hunk 9 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: This is not showing up in my pep8 command line check. Any idea why ?

File: benchmarks/benchmarks/stats.py (hunk 8 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: Oops, my bad. As they were params and initially I wasn't sure how they work , wrote separate blocks. Changed now. Thanks.

**Comment by maniteja123** [comment]
Now after doing the above changes, I tried to run benchmarks locally, I am getting this error. Not sure if something is wrong with my build or some break in code in master itself. Occurs after rebase also.
The traceback is https://gist.github.com/maniteja123/ca2a7ac732240c8b731d

File: benchmarks/benchmarks/stats.py (hunk 9 above)
Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: That's related to the pep8 error codes that are excluded in ``tox.ini`` in the root dir of the repo. If you copy this file to some temporary directory outside the source tree and directly run ``pep8`` on it, you'll get more violations.

**Comment by rgommers** [comment]
Allright, looks good and works fine for me. In it goes. Thanks @maniteja123!

File: benchmarks/benchmarks/stats.py (hunk 9 above)
Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: Oh, didn't know that. Thanks for the info. I suppose the following change now adheres to PEP8 ?
 
    a = np.random.rand(2,2) * 10

Comment (rgommers) [inline benchmarks/benchmarks/stats.py]: yes, that looks better

Comment (maniteja123) [inline benchmarks/benchmarks/stats.py]: Sorry, didn't see the update below ;) Thanks.

**Comment by maniteja123** [comment]
Oh great, thanks. Any insight into the error I am getting in my local build ?

**Comment by rgommers** [comment]
yes, I commented on your gist

**Comment by maniteja123** [comment]
@rgommers Thanks again :)

**Pull request merged**